### PR TITLE
Feature/dem 805/implement sampling rate float parameter

### PR DESCRIPTION
**Pull request by qSaevar**
Changes proposed in this pull request:
- Set and Get the sampling rate of the ZIUHFLI using float. 

@WilliamHPNielsen @jenshnielsen @astafan8

### CODE REVIEW

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 1]
```diff
@@ -1330,6 +1330,20 @@ def __init__(self, name: str, device_ID: str, **kwargs) -> None:
                             vals=vals.Enum(*list(self._samplingrate_codes.keys()))
                             )
 
+        self.add_parameter('scope_samplingrate_float',
+                           label="Scope's sampling rate as float",
+                           set_cmd=partial(self._set_samplingrate_as_float,
+                                           self._samplingrate_codes),
+                           unit='Hz',
+                           get_cmd=partial(self._get_samplingrate_as_float),
```
Comment (astafan8) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: no need for partial here

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 2]
```diff
@@ -2168,6 +2182,23 @@ def getduration():
...
+        converter = {'hz': 'e0', 'khz': 'e3', 'mhz': 'e6', 'ghz': 'e9',
+                     'thz': 'e12'}
+        value, suffix = frequency.split(' ')
+        return float(''.join([value, converter[suffix.lower()]]))
+
+    def _set_samplingrate_as_float(self, samplingrate_codes, frequency):
+        nearest_frequency = min(samplingrate_codes.keys(), key=lambda k: abs(
+            frequency - ZIUHFLI._convert_to_float(k)))
+
+        self.scope_samplingrate(nearest_frequency)
```
Comment (astafan8) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: smth like `self.scope_samplingrate_float._save_val(ZIUHFLI._convert_to_float(nearest_frequency))` is missing here, right, @QCoDeS/core ?

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py [hunk 3]
```diff
@@ -1330,6 +1330,20 @@ def __init__(self, name: str, device_ID: str, **kwargs) -> None:
                             vals=vals.Enum(*list(self._samplingrate_codes.keys()))
                             )
 
+        self.add_parameter('scope_samplingrate_float',
+                           label="Scope's sampling rate as float",
+                           set_cmd=partial(self._set_samplingrate_as_float,
+                                           self._samplingrate_codes),
+                           unit='Hz',
+                           get_cmd=partial(self._get_samplingrate_as_float),
+                           docstring=""" Set the scope sampling rate with float.
```
Comment (astafan8) [inline qcodes/instrument_drivers/ZI/ZIUHFLI.py]: i personally think the word "float" is abused in this PR. smth like "numeric" or "number" seems to be better. but i leave this to the author.

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 1 above)
Comment (qSaevar) [reply qcodes/instrument_drivers/ZI/ZIUHFLI.py]: ✔️

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 2 above)
Comment (qSaevar) [reply qcodes/instrument_drivers/ZI/ZIUHFLI.py]: right?

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 3 above)
Comment (qSaevar) [reply qcodes/instrument_drivers/ZI/ZIUHFLI.py]: ✔️

**Comment by qSaevar** [comment]
@astafan8 I addressed the your comments except the one with `self.scope_samplingrate_float._save_val(ZIUHFLI._convert_to_float(nearest_frequency))` as it was left with a question mark. Should I do it?

File: qcodes/instrument_drivers/ZI/ZIUHFLI.py (hunk 2 above)
Comment (astafan8) [reply qcodes/instrument_drivers/ZI/ZIUHFLI.py]: @QCoDeS/core ? :)

Comment (WilliamHPNielsen) [reply qcodes/instrument_drivers/ZI/ZIUHFLI.py]: See review below.

**Comment by qSaevar** [comment]
@WilliamHPNielsen I very well see your point and I'm not going to try and convince you. This behavior of setting the nearest point is how it works in the m4i driver, so there was a request for it in the uhfli driver as well.
But having the option of excepting and returning the sampling rate as float would still be very handy to have in many cases. So what if I make the "nearest" frequency a static method that finds the nearest allowed frequency, given some floating point number, that can in turn be used to set the sampling frequency? Something like:
```
sampling_rate = ZIUHFLI.round_to_nearest_sampling_frequency(desired_sampling_rate)
uhfli.scope_samplingrate_float(sampling_rate)
```
?

**Comment by WilliamHPNielsen** [comment]
@qSaevar I think what you propose with
```python
sampling_rate = ZIUHFLI.round_to_nearest_sampling_frequency(desired_sampling_rate)
uhfli.scope_samplingrate_float(sampling_rate)
```
is a nice and clean solution. 

Then the `scope_samplingrate_float` parameter can have an `Enum` validator only accepting the correct values, and the `round_to_nearest_sampling_frequency` method should then take care not to introduce floating-point errors, but that should not be too hard to implement.

**Comment by qSaevar** [comment]
@WilliamHPNielsen I updated the PR so now the `scope_samplingrate_float` parameter only accepts sampling rates that are allowed by the device. There is also a helping method that returns the closest sampling rate from some desired value and it is no longer linear as before.

**Comment by WilliamHPNielsen** [comment]
Looks terrific!

**Pull request merged**